userspace/libscap/engine/savefile/scap_reader_buffered.c

```diff
@@ -0,0 +1,147 @@
+} reader_handle_t;
+
+static int buffered_read(scap_reader_t *r, void* buf, uint32_t len)
+{
+    ASSERT(r != NULL);
+    reader_handle_t* h = (reader_handle_t*) r->handle;
+    uint8_t* buf_bytes = (uint8_t*) buf;
+    uint32_t size = 0;
+    while (len > 0 && !h->m_has_err)
+    {
+        if (h->m_buffer_off >= h->m_buffer_len)
+        {
+            int nread = h->m_reader->read(h->m_reader, h->m_buffer, h->m_buffer_cap);
+            if (nread <= 0)
+            {
+                // invalidate next read
+                h->m_has_err = true;
+                return buf_bytes - (uint8_t*) buf;
+            }
+            h->m_offset += nread;
+            h->m_buffer_off = 0;
+            h->m_buffer_len = (uint32_t) nread;
+        }
+        size = len < (h->m_buffer_len - h->m_buffer_off) ? len : (h->m_buffer_len - h->m_buffer_off);
+        memcpy(buf_bytes, h->m_buffer + h->m_buffer_off, size);
+        buf_bytes += size;
+        h->m_buffer_off += size;
+        len -= size;
+    }
+    return buf_bytes - (uint8_t*) buf;
+}
+
+static int64_t buffered_offset(scap_reader_t *r)
+{
+    ASSERT(r != NULL);
+    reader_handle_t* h = (reader_handle_t*) r->handle;
+    return h->m_offset;
+}
+
+static int64_t buffered_tell(scap_reader_t *r)
+{
+    ASSERT(r != NULL);
+    reader_handle_t* h = (reader_handle_t*) r->handle;
+    return h->m_offset - h->m_buffer_len + h->m_buffer_off;
+}
+
+static int64_t buffered_seek(scap_reader_t *r, int64_t offset, int whence)
+{
+    ASSERT(r != NULL);
+    reader_handle_t* h = (reader_handle_t*) r->handle;
+    if (whence == SEEK_CUR)
+    {
+        if (offset < 0 && h->m_buffer_off >= (uint32_t) (offset * -1))
+        {
+            h->m_buffer_off -= (uint32_t) (offset * -1);
+            return r->tell(r);
+        }
+        else if (offset > 0 && h->m_buffer_len >= h->m_buffer_off + (uint32_t) offset)
+        {
+            h->m_buffer_off += (uint32_t) offset;
+            return r->tell(r);
+        }
+    }
+    h->m_buffer_off = 0;
+    h->m_buffer_len = 0;
+    h->m_offset = h->m_reader->seek(h->m_reader, offset, whence);
+    return h->m_offset;
+}
+
+static const char* buffered_error(scap_reader_t *r, int *errnum)
+{
+    ASSERT(r != NULL);
+    reader_handle_t* h = (reader_handle_t*) r->handle;
+    return h->m_reader->error(h->m_reader, errnum);
+}
+
+static int buffered_close(scap_reader_t *r)
+{
+    ASSERT(r != NULL);
+    reader_handle_t* h = (reader_handle_t*) r->handle;
+    int res = h->m_reader->close(h->m_reader);
+    if (h->m_free_reader)
+    {
+        free(h->m_reader);
+    }
+    free(h->m_buffer);
+    free(r->handle);
+    return res;
+}
+
+scap_reader_t *scap_reader_open_buffered(scap_reader_t* reader, uint32_t bufsize, bool own_reader)
+{
+    if (reader == NULL || bufsize == 0)
+    {
+        return NULL;
+    }
+
+    reader_handle_t* h = (reader_handle_t *) malloc (sizeof (reader_handle_t));
```

Comment:
I would consider using `calloc` here, I think it would simplify the initialization of `h`.